walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -39,46 +41,90 @@
                 }
             }
         }
+        List<TimeRange> availableTimes = getAvailableSlots(bookedTimes, request.getDuration());
+ 
+        //Get blocked and available times for optional people
+        for(Event event: events){
+            for(String attendee: request.getOptionalAttendees()){
+                if(event.getAttendees().contains(attendee)){
+                    bookedOptionalTimes.add(event.getWhen());
+                    break;
+                }
+            }
+        }
+        List<TimeRange> availableOptionalTimes = getAvailableSlots(bookedOptionalTimes, request.getDuration());
 
-        if(bookedTimes.size() == 0){
-            slots.add(TimeRange.WHOLE_DAY);
+        // If no required attendee is available, then return no time slot available.
+        if(availableTimes.size() == 0){
             return slots;
         }
-
-        //Sort the booked times by start of day
-        Collections.sort(bookedTimes, TimeRange.ORDER_BY_START);
-
-        //Merge overlapping or nested timeslots into a single chunk of time
-        int i = 0;
-        while(i < bookedTimes.size() - 1){
-            if(bookedTimes.get(i).overlaps(bookedTimes.get(i+1)) || bookedTimes.get(i).contains(bookedTimes.get(i+1))){
-                int endTime = Math.max(bookedTimes.get(i).end(), bookedTimes.get(i+1).end());
-                bookedTimes.set(i, TimeRange.fromStartEnd(bookedTimes.get(i).start(), endTime, false));
-                bookedTimes.remove(i+1);
-            }else{
-                i++;
+ 
+        // Compare the timeslots available for required attendees and optional attendees and return the intersection.
+        for(int i = 0; i < availableTimes.size(); i++){
+            for(int j = 0; j < availableOptionalTimes.size(); j++){
+                if(availableTimes.get(i).end() < availableOptionalTimes.get(j).start()){
+                    break;
+                }
+                else if(availableTimes.get(i).equals(availableOptionalTimes.get(j))){
```

Comment:
do the "equals: and "contains" checks give different results for time intervals that are identical? In other words, can't you just have "contains" checks?

Reply:
You're right! I think contains includes conditions where they are equal. Will fix

Reply:
Resolved